### PR TITLE
[lldb] Add isConstant mode for FA locations

### PR DESCRIPTION
**Pull request by felipepiovezan**
This is similar to 9fe455fd0c7d, but for FA locations instead of register locations.
This is useful for unwind plans that cannot create abstract unwind rules, but instead must inspect the state of the program to determine the current CFA.

### CODE REVIEW

File: lldb/source/Target/RegisterContextUnwind.cpp [hunk 1]
```diff
@@ -2088,6 +2088,12 @@ bool RegisterContextUnwind::ReadFrameAddress(
     UnwindLogMsg("No suitable CFA found");
     break;
   }
+  case UnwindPlan::Row::FAValue::isConstant: {
+    address = fa.GetConstant();
+    address = abi_sp->FixCodeAddress(address);
```
Comment (felipepiovezan) [inline lldb/source/Target/RegisterContextUnwind.cpp:2093]: @jasonmolenda  I see a few other places in this file calling the function above (`FixCodeAddress`). But there are many other places that just take addresses and use them as is to do things like `Process:ReadMemory`... it makes me wonder whether we should audit those uses.

File: lldb/source/Target/RegisterContextUnwind.cpp [hunk 2]
```diff
@@ -2088,6 +2088,13 @@ bool RegisterContextUnwind::ReadFrameAddress(
     UnwindLogMsg("No suitable CFA found");
     break;
   }
+  case UnwindPlan::Row::FAValue::isConstant: {
+    address = fa.GetConstant();
+    if (abi_sp)
+      address = abi_sp->FixCodeAddress(address);
```
Comment (jasonmolenda) [inline lldb/source/Target/RegisterContextUnwind.cpp:2094]: We prefer people call `Process::FixDataAddress()` these days, a method which gets the ABI and runs the method.  (this is the address on the stack, so FixDataAddress is the correct method).  Yeah the code above at line 2051 should be doing the same thing, I know you were only copying it here.

**Review by jasonmolenda — APPROVED** [review]
LGTM.  Yes, the question of "when do we need to run an address through a Fix method" is largely driven by the different ABIs we encounter on different platforms.  If we know it is a virtual address value (no metadata in high bits/unaddressable bits that we may need to inspect), it is safe to run it through a Fix method.  The question of FixCode or FixData doesn't really matter here with aligned stack addresses, but a typical difference is that FixCode may clear the low 1 or 2 bits on targets with fixed length instructions, whereas FixData can point to a byte boundary so all of the low bits are typically preserved.

This does mean that we see Fix calls in generic code that are specific to how, e.g. Darwin's ABI signs a vtable pointer which might not be the case on all target using AArch64 ptrauth signing.  An example of where over-Fixing is a problem comes from Objective-C which has an "isa" pointer at the start of the object, identifying the object's metadata, and in some cases there may be a "non-pointer isa" (aka tagged pointer) which has the value of the object in the isa itself, e.g. a small NSNumber might be representable entirely within the isa word.  And so treating this non-pointer isa as a virtual address will clear significant bits and make it meaningless.

Practically speaking, we've moved to a model where we Fix addresses before we Read or Write from them in the Process, so even if we miss a Fix that was needed for a signed address, by the time we actually try to read from that address it will work.  This behavior is needed in the face of AArch64 MTE where we can have tags in the top nibble, or an AArch64 processor running in Top Byte Ignore mode where the inferior may store metadata in the type byte (Darwin's application processors run in TBI mode enabled).

File: lldb/source/Target/RegisterContextUnwind.cpp (hunk 2 above)
Comment (felipepiovezan) [reply lldb/source/Target/RegisterContextUnwind.cpp:2094]: Sounds good! I will push an updated version calling the process method!

**Pull request merged**